### PR TITLE
Add  vertical ordering to blog layouts

### PR DESCRIPTION
**Pull request by chmst**
Pull Request for Issue # .

### Summary of Changes
Proof of Concept 
There are many complaints of users who don't know how to add modifier classes for Blog Layouts. 
This PR is a proof of concept where the main classes for the modifiers are added to the blog layout params. 
I would like to have some feedback and if it is accepted can add it to featured. 

Open questions are, for example:
- Is it self explaining and helpful? 
- Do we need global params for that ?

### Testing Instructions
You need several articles in a category.

Now make a menu item for a category blog. In the Tab Blog Layout There are two new selections: Default classes. 

grafik
One for leading articles (the screenshot), the same for intro articles.

If the user selects "yes", the shown presents new input fields: 

grafik
The same for intro articles

Select here what you want, play with different combinations and see what happens with your blog. For example this: 

grafik

### Actual result BEFORE applying this Pull Request
There is one input field for the classes of the leading articles and one for the intro articles. 

### Expected result AFTER applying this Pull Request
There is one input field for the classes of the leading articles and one for the intro articles. 
There are input fields for default classes.

### Documentation Changes Required
yes

### CODE REVIEW

**Comment by brianteeman** [comment]
Testing this right now as this functionality is essential - thanks

> Do we need global params for that ?

yes definitely

**Comment by brianteeman** [comment]
1. We really need to resolve the three column layout thing. This page in particular is impossible to navigate. I had to hack it to one column to understand.

2.  I would move the #leading article to be the first field 
image

3.  As above with #intro articles

4. # columns should not be hidden

5. Blog Class Default Classes
Is this just a label to expand the other fields? If it is then it should be a switcher not a select and needs a better title

This is AWESOME

File: administrator/language/en-GB/joomla.ini [hunk 1]
```diff
@@ -352,8 +352,16 @@ JGLOBAL_BATCH_MOVE_ROW_NOT_FOUND="Can't find the destination row for this move."
 JGLOBAL_BATCH_PROCESS="Process"
 JGLOBAL_BATCH_WORKFLOW_STATE_ROW_NOT_FOUND="Can't find the destination row for this state change."
 JGLOBAL_BLOG="Blog"
-JGLOBAL_BLOG_CLASS="Blog Class"
+JGLOBAL_BLOG_CLASS_BOXED="Boxed Layout"
+JGLOBAL_BLOG_CLASS_COLUMNS="Number of Columns"
+JGLOBAL_BLOG_CLASS_HELP="Blog Class Default Classes"
+JGLOBAL_BLOG_CLASS_IMAGE="Image Alingment"
```
Comment (brianteeman) [inline administrator/language/en-GB/joomla.ini]:
```suggestion
JGLOBAL_BLOG_CLASS_IMAGE="Image Alignment"
```

**Comment by SharkyKZ** [comment]
These classes are specific to Cassiopeia and should not be hardcoded in `com_content`.

**Comment by chmst** [comment]
@brianteeman changes made as suggested.

@SharkyKZ have you an idea how we could make this possible? Maybe a plugin in cassiopeia?

**Comment by brianteeman** [comment]
@chmst thanks. Will test further

Regarding @sharkyz comment. I dont see this PR as any different to j3 and its use of bootstrap classes.

This PR is essential if we dont want to make it way too hard for users to upgrade successfully or to use existing functionality. Right now columns etc dont do anything so users think it is broken

**Comment by chmst** [comment]
I understand that this is specific for cassiopeia. But it is hard to make a strict separation between core and template. Also the difference of leading items and blog items can be seen as template specific, or the intro images vs. fulltext images.
The changes in this PR are rather compact and they don't introduce dependencies so I think that it is an advantage for acceptance. <hr /><sub>This comment was created with the <a href="https://github.com/joomla/jissues">J!Tracker Application</a> at <a href="https://issues.joomla.org/tracker/joomla-cms/30910">issues.joomla.org/tracker/joomla-cms/30910</a>.</sub>

**Comment by ceford** [comment]
On this patch: my intro image came out very small, about the same size as a capital M. The other images were OK. To me it seems crazy that the 1/n column layout for Leading and Intro articles is not available by default. I remembered it from somewhere (J3?) and spent ages looking for it.

screen shot 2020-10-04 at 13 22 20<hr /><sub>This comment was created with the <a href="https://github.com/joomla/jissues">J!Tracker Application</a> at <a href="https://issues.joomla.org/tracker/joomla-cms/30910">issues.joomla.org/tracker/joomla-cms/30910</a>.</sub>

**Comment by ceford** [comment]
I should have said also: I think the images are in the wrong place! I would float the image left or right of the text and add some padding.<hr /><sub>This comment was created with the <a href="https://github.com/joomla/jissues">J!Tracker Application</a> at <a href="https://issues.joomla.org/tracker/joomla-cms/30910">issues.joomla.org/tracker/joomla-cms/30910</a>.</sub>

**Comment by ChristineWk** [comment]
Maybe I missed something?
screen shot 2020-10-04 at 20 59 30

Can't find Boxed Layout for columns and don't know what the Default Classes are. Sorry.<hr /><sub>This comment was created with the <a href="https://github.com/joomla/jissues">J!Tracker Application</a> at <a href="https://issues.joomla.org/tracker/joomla-cms/30910">issues.joomla.org/tracker/joomla-cms/30910</a>.</sub>

**Comment by ReLater** [comment]
I don't know if it's of interest for this pr and Joomla 4 but as far as I know the **CSS based** "masonry" layouts (cards) have been removed from Bootytrap 5 for good reasons ("strange ordering and things"). One should not expect that it has a future for coming templates.
https://github.com/twbs/bootstrap/pull/28922

**Comment by gostn** [comment]
@ChristineWk 

image

**Comment by chmst** [comment]
Thank you all for testing and your meaning. We are working on a better solution in cassiopiea. I cles this for now.

**Comment by ChristineWk** [comment]
@gostn Thank you for your image. It's clear now. It belongs to "Fields are expanded".

**Comment by chmst** [comment]
Re-Opening this as it can resolve the release blocker #27478  
Needs calculationg the ordering number in blog layout.

vertical-blog

**Comment by brianteeman** [comment]
Is vertical ordering the same as the J3 option Multi Column Ordering?
image

If it is then can you rename it here.

**Comment by ChristineWk** [comment]
I have tested this item :white_check_mark: successfully on dde166264011a74990b967a30c38460d37d5528f<hr /><sub>This comment was created with the <a href="https://github.com/joomla/jissues">J!Tracker Application</a> at <a href="https://issues.joomla.org/tracker/joomla-cms/30910">issues.joomla.org/tracker/joomla-cms/30910</a>.</sub>

**Comment by ghost** [comment]
I have tested this item :white_check_mark: successfully on dde166264011a74990b967a30c38460d37d5528f<hr /><sub>This comment was created with the <a href="https://github.com/joomla/jissues">J!Tracker Application</a> at <a href="https://issues.joomla.org/tracker/joomla-cms/30910">issues.joomla.org/tracker/joomla-cms/30910</a>.</sub>

**Comment by Harmageddon** [comment]
Should the vertical ordering parameter have any effect yet? Because I don't see any, and I can't find where the parameter is being used in the view.

(Edit: Didn't want to pressure here, as the PR is still a draft. I was just wondering, but assume it's not yet implemented. ;-) )

**Comment by ghost** [comment]
who care on a draft having 2 successfully test

**Comment by richard67** [comment]
@gostn It's the author who decides if a PR is draft or not.

**Comment by ghost** [comment]
@chmst what now, test are wastet time?

**Comment by HLeithner** [comment]
@gostn testing a draft may makes sense but mainly for discussion. If the PR didn't get changed the tests are still valid.

**Comment by ghost** [comment]
@HLeithner thanks for information not like https://github.com/joomla/joomla-cms/pull/30716#issuecomment-736284377 by @richard67

**Comment by chmst** [comment]
As it is easier for migration I have restored the two params from J3 for ordering down and number of columns.   If a user chooses the order down, the input in blog class is ignored and the into items are displaed in columns.

**Comment by ChristineWk** [comment]
When using this:

screen shot 2020-12-01 at 19 13 56
it looks OK.

When using this:
screen shot 2020-12-01 at 19 14 36

Frontend:
Notice: Undefined variable: numCols in /home/xyz/components/com_content/tmpl/category/blog.php on line 102 Notice: Undefined variable: numCols in /home/xyz/components/com_content/tmpl/category/blog.php on line 102 
<hr /><sub>This comment was created with the <a href="https://github.com/joomla/jissues">J!Tracker Application</a> at <a href="https://issues.joomla.org/tracker/joomla-cms/30910">issues.joomla.org/tracker/joomla-cms/30910</a>.</sub>

**Comment by chmst** [comment]
Thank you @ChristineWk, fixed

**Comment by brianteeman** [comment]
Instead of Ordering option down can you use the J3 terminology

https://github.com/joomla/joomla-cms/pull/30910#issuecomment-734746368

**Comment by chmst** [comment]
@brianteeman I think, it is not quite the same because this ordering is ONLY for vertical ordering (down).  He cannot choose fixed number of columns ad ordering "across" as before, this is made via modifier classes. 

This solution is not really good, but at least enables a migration from J3 to j4. 

An open question is: Is this option for leading articles and intro-articles? At the moment it is for intro-articles only.

**Comment by ChristineWk** [comment]
@chmst Thank you. Notice is fixed.

**Comment by Harmageddon** [comment]
Hmm, it kind of works for me, but now the value of "Blog Class (Intro Articles)" is ignored completely when using vertical layout. I think that's not optimal either. Of course, we get a conflict if someone enters "columns-3" for the class and then vertical layout with 2 columns. But what if someone wants to use any other class for the intro articles?

**Comment by ChristineWk** [comment]
@Harmageddon Thks for your investigation :-)
> 
> 
> Hmm, it kind of works for me, but now the value of "Blog Class (Intro Articles)" is ignored completely when using vertical layout. I think that's not optimal either. Of course, we get a conflict if someone enters "columns-3" for the class and then vertical layout with 2 columns. But what if someone wants to use any other class for the intro articles?

Maybe it belongs to this?


> 
> 
> If a user chooses the order down, the input in blog class is ignored and the into items are displaed in columns.

**Comment by Harmageddon** [comment]
> Maybe it belongs to this?
> 
> > If a user chooses the order down, the input in blog class is ignored and the into items are displaed in columns.

Ah yes, thank you. Still, I think it's not optimal, because the blog class can be used for other things than just columns. Maybe do a search/replace and remove the "columns-X" part in this case? In any case, this behaviour should be documented somewhere, because users will wonder why their input is ignored.

**Comment by Harmageddon** [comment]
I'm trying to tackle this problem from a different angle, by providing an update script that converts the old column parameters to the according CSS classes. PR should be ready tomorrow or friday. ;-)

**Comment by paritshivani** [comment]
I have tested this item :red_circle: unsuccessfully on 39cbad47bd6b9fed645873eeafcab34dab1b4ea4<hr /><sub>This comment was created with the <a href="https://github.com/joomla/jissues">J!Tracker Application</a> at <a href="https://issues.joomla.org/tracker/joomla-cms/30910">issues.joomla.org/tracker/joomla-cms/30910</a>.</sub>

**Comment by paritshivani** [comment]
I have tested this item :red_circle: unsuccessfully on 39cbad47bd6b9fed645873eeafcab34dab1b4ea4<hr /><sub>This comment was created with the <a href="https://github.com/joomla/jissues">J!Tracker Application</a> at <a href="https://issues.joomla.org/tracker/joomla-cms/30910">issues.joomla.org/tracker/joomla-cms/30910</a>.</sub>

**Comment by ChristineWk** [comment]
Vertical and horizontal arrangement look OK with me. But I don't know if this is enough for this PR. See also comments & new PR.

Vertical:
screen shot 2020-12-05 at 12 02 33

Horizontal:
screen shot 2020-12-05 at 12 02 54<hr /><sub>This comment was created with the <a href="https://github.com/joomla/jissues">J!Tracker Application</a> at <a href="https://issues.joomla.org/tracker/joomla-cms/30910">issues.joomla.org/tracker/joomla-cms/30910</a>.</sub>

**Comment by Tejashrimajage** [comment]
I have tested this item :red_circle: unsuccessfully on 39cbad47bd6b9fed645873eeafcab34dab1b4ea4<hr /><sub>This comment was created with the <a href="https://github.com/joomla/jissues">J!Tracker Application</a> at <a href="https://issues.joomla.org/tracker/joomla-cms/30910">issues.joomla.org/tracker/joomla-cms/30910</a>.</sub>

**Comment by chmst** [comment]
Closing this in favour to #31570

**Comment by chmst** [comment]
Reopening as a possible solution for issue #32012 - must be reviewed.

This solution could be better for users especially if the three-columns layout is set back and the blocks for intro and blog fields can be separated clearly.

**Comment by richard67** [comment]
@chmst Will this still work now as #31570 has been merged?

**Comment by chmst** [comment]
I think yes. But as it must be reviewed.We have enough to do, so I close again and  we can think again this in  one of the following versions.

**Comment by Harmageddon** [comment]
Hmm... #31570 converts the column parameters into CSS classes. If we want to re-introduce column parameters or some other sort of display parameters, it might be better to revert it, because the information about columns will be "lost" on upgraded sites (not lost, but transferred to the column-X class).
For new sites, #31570 doesn't change anything. But for sites that are updated from 3.x to 4.0, I think it would be better to make the decision now, whether to go only with CSS classes or with more sophisticated parameters.

**Pull request closed without merging**